---
title: About
layout: page
---
![Profile Image]({{ site.url }}/{{ site.picture }})

I am currently a graduate student at [Virginia Tech](http://www.vt.edu/), in the Genetics, Bioinformatics, and Computational Biology Program, and my current research interest involves dynamical modeling of biological processes, specifically in the context of yeast biology.
I am currently co-advised by Professors [John Tyson](http://mpf.biol.vt.edu/lab_website/) and [T.M. Murali](http://bioinformatics.cs.vt.edu/~murali/). Details of my current research project can be found [here]({{site.url}}/nutrient-signaling).

Previously, in Fall 2016, I worked with [Dr. Silke Hauf](http://www.hauflab.org/) during my graduate school rotations, where I was investigating the phenomenon of *order in chromosome segregation* in *S. pombe*; simultaneously, I was developing an automated image analysis work-flow to analyse dynamics of cell-cycle regulators.

Before I joined Virginia Tech, in Spring 2015, I worked with [Dr. Abby Dernburg](https://mcb.berkeley.edu/labs/dernburg/) at [UC Berkeley](http://www.berkeley.edu/) where I was involved in aproject which combined comparative evolution and meiosis, focussing on the variation in meiotic recombination machinery in the model nematode *C. elegans* and its cousin *P. pacificus*. There, I spent most of my time optimizing imaging protocols for fixed cell fluorescence imaging in  *P. pacifius*.

During the Summer of 2014, as part of [NIGINTERN](https://www.nig.ac.jp/jimu/soken/intern/2017/index.html) program at the [National Institute of Genetics, Japan](https://www.nig.ac.jp/nig/), I carried out a small project under the guidance of [Dr. Akatsuki Kimura](https://www.nig.ac.jp/nig/research/organization-top/organization/kimura) where I was introduced to nematode biology and image analysis.

This webpage is my attempt at documenting some of my computational adventures systematically, and is essentially a hodge-podge of random things I have learned in my journey as a PhD student so far. Apart from microscopy, scripting and mathematical modeling, I also like listening to and making music.

# Publications and Preprints

- Pratapa, A., **Jalihal, A. P.**, Law, J. N., Bharadwaj, A., and
  Murali, T.M., (2019) Benchmarking algorithms for gene regulatory
  network inference from single-cell transcriptomic data. [preprint](https://doi.org/10.1101/642926), [Beeline](https://github.com/Murali-group/Beeline), [BoolODE](https://github.com/Murali-group/BoolODE)
- Pratapa, A., **Jalihal, A. P.**, Ravi, S. S., and Murali, T. M. (2018,
  August). Efficient Synthesis of Mutants Using Genetic Crosses. In
  Proceedings of the 2018 ACM International Conference on
  Bioinformatics, Computational Biology, and Health Informatics
  (pp. 53-62). ACM. [doi](https://doi.org/10.1145/3233547.3233556)

